coderd/prometheusmetrics/aggregator.go

```diff
@@ -146,13 +147,40 @@ func (ma *MetricsAggregator) Run(ctx context.Context) func() {
 
 				output := make([]prometheus.Metric, 0, len(ma.queue))
 				for _, m := range ma.queue {
-					desc := prometheus.NewDesc(m.Name, metricHelpForAgent, agentMetricsLabels, nil)
+					labels := make([]string, len(agentMetricsLabels)+len(m.Labels))
+					var labelIndex int
+					for _, l := range agentMetricsLabels {
+						labels[labelIndex] = l
+						labelIndex++
+					}
+
+					for _, l := range m.Labels {
+						labels[labelIndex] = l.Name
+						labelIndex++
+					}
+
+					desc := prometheus.NewDesc(m.Name, metricHelpForAgent, labels, nil)
 					valueType, err := asPrometheusValueType(m.Type)
 					if err != nil {
 						ma.log.Error(ctx, "can't convert Prometheus value type", slog.F("name", m.Name), slog.F("type", m.Type), slog.F("value", m.Value), slog.Error(err))
 						continue
 					}
-					constMetric := prometheus.MustNewConstMetric(desc, valueType, m.Value, m.username, m.workspaceName, m.agentName)
+
+					labelValues := make([]string, len(agentMetricsLabels)+len(m.Labels))
+					labelValues[0] = m.username
+					labelValues[1] = m.workspaceName
+					labelValues[2] = m.agentName
+
+					if len(m.Labels) > 0 {
+						labelIndex = 3
+
+						for _, l := range m.Labels {
+							labelValues[labelIndex] = l.Value
+							labelIndex++
+						}
+					}
+
+					constMetric := prometheus.MustNewConstMetric(desc, valueType, m.Value, labelValues...)
```

Comment:
I'd suggest pulling this entire procedure out into a method on `annotatedMetric`, eg

```
constMetric, err := m.asPrometheus()
```

Reply:
Good idea 👍